src/typing/env_js.ml

```diff
@@ -661,24 +771,21 @@ let merge_env =
           (string_of_reason reason)
           name)
       (* global lookups may leave new entries in one or both child envs *)
-      | None, child1, child2 when env0 = [] ->
+      | None, _, _ when env0_ = [] ->
         (* ...in which case we can forget them *)
         ()
-      (* otherwise, non-refinement uneven distributions are asserts. *)
-      (* TODO:
-         Asserting on all uneven distributions is too strict for now.
-         let_env can interact with path-dependent refinement to leave
-         residual entries in child contexts, when catch blocks combine
-         conditionals and early returns.
-         So for now we need to let extra child entries pass, though we
-         should still assert on extra parent entries.
-         Once let/const/LexScope is in place, we can reimplement let_env
-         in terms of those constructs instead, and the full assert can
-         be reinstated.
-       *)
-      | None, _, _ ->
+      (* changeset entry exists only in lex scope *)
+      | None, None, Some e when is_lex scope2 && Entry.is_lex e ->
+        ()
+      | None, Some e, None when is_lex scope1 && Entry.is_lex e ->
         ()
-      | Some _ as orig, child1, child2 ->
+      (* walk through uneven lex scopes *)
```

Comment:
Just to be clear: I'm saying I don't think the lists will ever be ragged, *if* `merge_env` is called properly, i.e. I think it's a precondition of calling `merge_env` that you've come out of any divergent lexical scopes. And I think we should capture that by asserting if we get scope lists that couldn't happen if the precondition is satisfied. (Rather than handling them.)